pkg/sql/sem/tree/schema_helpers.go

```diff
@@ -32,14 +32,51 @@ func IsSetOrResetSchemaLocked(n Statement) bool {
 // IsAllowedLDRSchemaChange returns true if the schema change statement is
 // allowed to occur while the table is being referenced by a logical data
 // replication job as a destination table.
-func IsAllowedLDRSchemaChange(n Statement) bool {
+func IsAllowedLDRSchemaChange(n Statement, virtualColNames []string) bool {
 	switch s := n.(type) {
 	case *CreateIndex:
-		// Only allow non-unique and non-partial indexes to be created. A unique or
-		// partial index on a destination table could cause inserts to fail.
-		return !s.Unique && s.Predicate == nil
+		// Don't allow creating an index on a virtual column.
+		for _, col := range s.Columns {
+			if slices.Contains(virtualColNames, string(col.Column)) {
+				return false
+			}
+		}
+		// Disallow unique, partial, or hash-sharded indexes. Having these indexes
+		// on a destination table could cause inserts to fail.
+		// NB: hash-sharded indexes are disallowed since they create an index on a
+		// virtual column. Since it also implicitly creates the virtual column
+		// at the same time, the check above on virtualColNames would not block it.
+		return !s.Unique && s.Predicate == nil && s.Sharded == nil
 	case *DropIndex:
 		return true
+	case *SetZoneConfig:
+		return true
+	case *AlterTable:
+		onlySafeStorageParams := true
+		for _, cmd := range s.Cmds {
+			switch c := cmd.(type) {
+			// Allow safe storage parameter changes.
+			case *AlterTableSetStorageParams:
+				// ttl_expire_after is not safe since it creates a new column and
+				// backfills it.
+				if c.StorageParams.GetVal("ttl_expire_after") != nil {
+					onlySafeStorageParams = false
+				}
+			case *AlterTableResetStorageParams:
+				if slices.Contains(c.Params, "ttl_expire_after") {
```

Comment:
not sure why this branch gets the params via `c.Params` while the above branch gets them via `c.StorageParams`. Do both fields need to be checked in both branches?

Reply:
it's just because the fields are named different things in the `AlterTableResetStorageParams` and `AlterTableSetStorageParams` structs:
https://github.com/cockroachdb/cockroach/blob/9f61a38b670e57873509568f67bb665e50cd1935/pkg/sql/sem/tree/alter_table.go#L619
https://github.com/cockroachdb/cockroach/blob/9f61a38b670e57873509568f67bb665e50cd1935/pkg/sql/sem/tree/alter_table.go#L636

Reply:
i see. thanks!